engine/src/cubos/engine/imgui/data_inspector.cpp

```diff
@@ -0,0 +1,309 @@
+}
+
+void DataInspector::showStructured(const std::string& name, const Type& type, auto fn)
+{
+    ImGui::TableSetColumnIndex(0);
+    ImGui::AlignTextToFramePadding();
+    bool nodeOpen = ImGui::TreeNode(name.c_str());
+    if (ImGui::IsItemHovered(ImGuiHoveredFlags_AllowWhenDisabled))
+    {
+        ImGui::SetTooltip("%s", type.name().c_str());
+    }
+
+    if (nodeOpen)
+    {
+        fn();
+        ImGui::TreePop();
+    }
+}
+
+#define GET_INPUT_SCALAR_IF_TYPE(TYPE, DATATYPE)                                                                       \
+    if (type.is<TYPE>())                                                                                               \
+    {                                                                                                                  \
+        if (ImGui::InputScalar(name.c_str(), DATATYPE, value))                                                         \
+        {                                                                                                              \
+            modified = true;                                                                                           \
+        }                                                                                                              \
+    }
+
+bool DataInspector::editPrimitive(const std::string& name, const Type& type, void* value)
+{
+    ImGui::TableSetColumnIndex(0);
+    ImGui::AlignTextToFramePadding();
+    ImGui::TreeNodeEx(name.c_str(),
+                      ImGuiTreeNodeFlags_Leaf | ImGuiTreeNodeFlags_NoTreePushOnOpen | ImGuiTreeNodeFlags_Bullet);
+    if (ImGui::IsItemHovered(ImGuiHoveredFlags_AllowWhenDisabled))
+    {
+        ImGui::SetTooltip("%s", type.name().c_str());
+    }
+
+    auto str = getPrimitiveValue(value, type);
+    ImGui::PushID(name.c_str());
+    ImGui::TableSetColumnIndex(1);
+    bool modified = false;
+    GET_INPUT_SCALAR_IF_TYPE(int8_t, ImGuiDataType_S8)
+    GET_INPUT_SCALAR_IF_TYPE(int16_t, ImGuiDataType_S16)
+    GET_INPUT_SCALAR_IF_TYPE(int32_t, ImGuiDataType_S32)
+    GET_INPUT_SCALAR_IF_TYPE(int64_t, ImGuiDataType_S64)
+    GET_INPUT_SCALAR_IF_TYPE(uint8_t, ImGuiDataType_U8)
+    GET_INPUT_SCALAR_IF_TYPE(uint16_t, ImGuiDataType_U16)
+    GET_INPUT_SCALAR_IF_TYPE(uint32_t, ImGuiDataType_U32)
+    GET_INPUT_SCALAR_IF_TYPE(uint64_t, ImGuiDataType_U64)
+    GET_INPUT_SCALAR_IF_TYPE(float, ImGuiDataType_Float)
+    GET_INPUT_SCALAR_IF_TYPE(double, ImGuiDataType_Double)
+    if (type.is<bool>())
+    {
+        if (ImGui::Checkbox(name.c_str(), (bool*)value))
+        {
+            modified = true;
+        }
+    }
+    if (type.is<std::string>())
+    {
+        if (ImGui::InputText(name.c_str(), &mTextBuffer, ImGuiInputTextFlags_EnterReturnsTrue))
+        {
+            auto* p = static_cast<std::string*>(value);
+            *p = mTextBuffer;
+            modified = true;
+        }
+    }
+    ImGui::PopID();
+
+    return modified;
```

Comment:
You should give an error and show ??? or something if the type is unsupported here, right now you just return without doing anything